Formal concrete relation (#47)
zkpstandard/wg-plonkish: src/optimizations.md

| $L_v$             | Number of table columns in the lookup table with index $v$, $\mathsf{TAB}_v$. | [Lookup constraints](#lookup-constraints) |
| $\mathsf{TAB}_v$  | Lookup tables $\mathsf{TAB}_v \subseteq \mathbb{F}^{L_v}$, each with a number of tuples in $\mathbb{F}^{L_v}$. | [Lookup constraints](#lookup-constraints) |
| $q_{v,s}$         | ✨ Scaling multivariate polynomials $q_{v,s} \mathrel{⦂} \mathbb{F}^{d  m} \rightarrow \mathbb{F}$ for $s \leftarrow 0 \text{..} L_v$. | [Lookup constraints](#lookup-constraints) |
| $\mathsf{LOOK}_v$ | Sets $\mathsf{LOOK}_v \subseteq [0,n)$ indicating rows on which the scaling polynomials $q_{v,s}$ evaluate to some tuple in $\mathsf{TAB}_v$. | [Lookup constraints](#lookup-constraints) |

Multivariate polynomials are defined below in the [Custom constraints](#custom-constraints) section.

### Witnesses

The relation $\mathcal{R}_{\mathsf{concrete}}$ takes witnesses of the following form:

| Witness element   | Description |
| ----------------- | -------- |
| $w$               | The witness matrix $w \mathrel{⦂} \mathbb{F}^{m \times n}$. |

> ✨ Define $\vec{w}_{j} \in \mathbb{F}^{m d}$ as the row vector $\big[\, w[i, j + \mathsf{offset}] : (i, \mathsf{offset}) \leftarrow (0 \text{..} m) \times \mathsf{offsets} \,\big]$.

Author: Should be an error to have a constrained entry which is out of range.

The optimizations suggested in this specification would be impractical or error-prone to perform manually, as they require global reasoning about the witness layout. Without automation, ensuring correctness while reordering rows or applying offsets becomes fragile and difficult to scale, especially in the presence of copy constraints or lookups where subtle errors can silently break circuit soundness.

In this document, we:

* **Define the concrete Plonkish relation**, intended for readers building a proving system or interactive oracle proof targeting the concrete Plonkish relation.

* **Describe a correctness-preserving circuit translation**, for those verifying the optimizations introduced in this document.

* **Specify the translation from an abstract to a concrete Plonkish relation**, aimed at implementers of the concrete relation.

* **Prove that the translation preserves correctness**, again intended for readers verifying the soundness of the optimisations.

## The Concrete Plonkish Relation

The relation $\mathcal{R}_{\mathsf{concrete}}$ is an optimisation of $\mathcal{R}_{\mathsf{plonkish}}$.  We have highlighted differences using the icon ✨.  

> 

Reviewer: ```suggestion
We will use the convention that variables marked with a prime ($'$) refer to *concrete* column or row indices.

```

| $m_f \leq m$      | Number of columns that are fixed. | [Fixed constraints](#fixed-constraints) |
| $f$               | The fixed content of the first $m_f$ columns, $f \mathrel{⦂} \mathbb{F}^{m_f \times n}$. | [Fixed constraints](#fixed-constraints) |
| $p_u$             | ✨ Custom multivariate polynomials $p_u \mathrel{⦂} \mathbb{F}^{d  m} \rightarrow \mathbb{F}$. | [Custom constraints](#custom-constraints) |
| $\mathsf{CUS}_u$  | Sets $\mathsf{CUS}_u \subseteq [0,n)$ indicating rows on which the custom polynomials $p_u$ are constrained to evaluate to 0. | [Custom constraints](#custom-constraints) |
| $L_v$             | Number of table columns in the lookup table with index $v$, $\mathsf{TAB}_v$. | [Lookup constraints](#lookup-constraints) |
| $\mathsf{TAB}_v$  | Lookup tables $\mathsf{TAB}_v \subseteq \mathbb{F}^{L_v}$, each with a number of tuples in $\mathbb{F}^{L_v}$. | [Lookup constraints](#lookup-constraints) |
| $q_{v,s}$         | ✨ Scaling multivariate polynomials $q_{v,s} \mathrel{⦂} \mathbb{F}^{d  m} \rightarrow \mathbb{F}$ for $s \leftarrow 0 \text{..} L_v$. | [Lookup constraints](#lookup-constraints) |

> | $\mathsf{LOOK}_v$ | Sets $\mathsf{LOOK}_v \subseteq [0,n)$ indicating rows on which the scaling polynomials $q_{v,s}$ evaluate to some tuple in $\mathsf{TAB}_v$. | [Lookup constraints](#lookup-constraints) |

Reviewer: ```suggestion
| $\mathsf{LOOK}_v$ | Sets $\mathsf{LOOK}_v \subseteq [0,n)$ indicating the source rows at which the scaling polynomials $q_{v,s}$ evaluate to some tuple in $\mathsf{TAB}_v$. | [Lookup constraints](#lookup-constraints) |
```

### Definition of the relation

Given the above definitions, the relation $\mathcal{R}_{\mathsf{concrete}}$ corresponds to a set of $\,(\!$ instance $\!,\,$ witness $\!)\,$ pairs
$$
 \left(x = \left(\mathbb{F}, C = \left(d, \mathsf{offsets}, t, n, m, \equiv, S, m_f, f, \left[\, (p_u, \mathsf{CUS}_{u}) \,\right]_u, \left[\, (L_v, \mathsf{TAB}_v, \left[\, q_{v,s} \,\right]_s, \mathsf{LOOK}_v) \,\right]_v\right), \phi\right),\, w \right)
$$
such that:
$$
\begin{array}{ll|l}
   w \mathrel{⦂} \mathbb{F}^{m \times n}, \ f \mathrel{⦂} \mathbb{F}^{m_f \times n} & & i \in [0,m_f), \ j \in [0,n) \Rightarrow w[i, j] = f[i, j] \\[0.3ex]
   S \subseteq ([0,m) \times [0,n)) \times [0,t), \ \phi \mathrel{⦂} \mathbb{F}^t & & ((i,j),k) \in S \Rightarrow w[i, j] = \phi[k] \\[0.3ex]
   \equiv\; \subseteq ([0,m) \times [0,n)) \times ([0,m) \times [0,n)) & & (i,j) \equiv (k,\ell) \Rightarrow w[i, j] = w[k, \ell] \\[0.3ex]

>    \mathsf{CUS}_u \subseteq [0,n), \ p_u \mathrel{⦂} \mathbb{F}^{d  m} \rightarrow \mathbb{F} & & j \in \mathsf{CUS}_u \Rightarrow p_u(\vec{w}_j) = 0 \\[0.3ex]
>    \mathsf{LOOK}_v \subseteq [0,n), \ q_{v,s} \mathrel{⦂} \mathbb{F}^{d  m} \rightarrow \mathbb{F}, \ \mathsf{TAB}_v \subseteq \mathbb{F}^{L_v} & & j \in \mathsf{LOOK}_v \Rightarrow \big[\, q_{v,s}(\vec{w}_j) : s \leftarrow 0 \text{..} L_v \,\big] \in \mathsf{TAB}_v

Reviewer: ```suggestion
   \mathsf{CUS}_u \subseteq [0,n), \ p_u \mathrel{⦂} \mathbb{F}^{m \cdot d} \rightarrow \mathbb{F} & & j \in \mathsf{CUS}_u \Rightarrow p_u(\vec{w}_j) = 0 \\[0.3ex]
   \mathsf{LOOK}_v \subseteq [0,n), \ q_{v,s} \mathrel{⦂} \mathbb{F}^{m \cdot d} \rightarrow \mathbb{F}, \ \mathsf{TAB}_v \subseteq \mathbb{F}^{L_v} & & j \in \mathsf{LOOK}_v \Rightarrow \big[\, q_{v,s}(\vec{w}_j) : s \leftarrow 0 \text{..} L_v \,\big] \in \mathsf{TAB}_v
```

There are four types of constraints that a Plonkish statement $(x, w) \in \mathcal{R}_{\mathsf{concrete}}$ must satisfy:

* Fixed constraints
* Copy constraints
* Custom constraints
* Lookup constraints

An Interactive Oracle Proof for an optimised Plonkish constraint system must demonstrate that all of these contraints are satisfied by $\,(\!$ instance $\!,\,$ witness $\!) \in \mathcal{R}_{\mathsf{concrete}}\,$

#### Fixed constraints

The first $m_f$ columns of $w$ are fixed to the columns of $f$.

#### Custom constraints

In the concrete model we define here, a custom constraint applies to a set of offset rows relative to each row selected for that constraint.

Custom constraints enforce that witness entries within a row satisfy some multivariate polynomial. Here $p_u$ could indicate any case that can be generated using a combination of multiplications and additions.

| Custom Constraints | Description |
| -------- | -------- | 
| $j \in \mathsf{CUS}_u \Rightarrow p_u(\vec{w}_j) = 0$ | $u$ is the index of a custom constraint. $j$ ranges over the set of rows $\mathsf{CUS}_u$ <br> for which the custom constraint is switched on. |

> Here $p_u \mathrel{⦂} \mathbb{F}^{d  m} \rightarrow \mathbb{F}$ is an arbitrary [multivariate polynomial](https://en.wikipedia.org/wiki/Polynomial_ring#Definition_(multivariate_case)):

Reviewer: ```suggestion
Here $p_u \mathrel{⦂} \mathbb{F}^{m \cdot d} \rightarrow \mathbb{F}$ is an arbitrary [multivariate polynomial](https://en.wikipedia.org/wiki/Polynomial_ring#Definition_(multivariate_case)):
```

> Given $\eta$ symbols $X_0, \dots, X_{\eta-1}$ called indeterminates, a multivariate polynomial $P$ in these indeterminates, with coefficients in $\mathbb{F}$,
> is a finite linear combination 
> 
> $P(X_0, \dots, X_{\eta-1}) = \sum_{z=0}^{\nu-1} \Big(c_z\, {\small\prod_{b=0}^{\eta-1}}\, X_b^{\alpha_{z,b}}\Big)$
> 
>  where $\nu \mathrel{⦂} \mathbb{N}$, $c_z \mathrel{⦂} \mathbb{F} \neq 0$, and $\alpha_{z,b} \mathrel{⦂} \mathbb{N}$.

> 

Reviewer: ```suggestion
Note that in this usage $\eta = m \cdot d$.

```

#### Lookup constraints

Lookup constraints enforce that some polynomial function of the witness entries on a row are contained in some table.

> In this specification, we only support fixed lookup tables determined in advance. This could be generalized to support dynamic tables determined by part of the witness matrix.

Reviewer: This is where I got up to in my review when pairing with @mmaller.

    +-------------------+             +-------------+
```

The function  
$$
\mathsf{compute\_coord\_map} 
$$  
takes as input an abstract circuit and a set of designer-provided hints $(C, \mathsf{hints}) \in \mathsf{AbstractCircuit} \times \mathsf{Hints}$.  The domain of abstract circuits, $\mathsf{AbstractCircuit}$, is defined formally in the description of the relation $\mathcal{R}_{\mathsf{plonkish}}$.  The domain of offset hints, $\mathsf{Hints}$, is defined in the following section.

It returns $(d', \mathsf{offsets}', m', n', \mathsf{coord\_map})$ such that :
- $d' \in \mathbb{N}$ is the number of offsets
- $\mathsf{offsets}' \in \mathbb{Z}^{d}$ are the offsets
- $m' \in \mathbb{N}$ is the number of concrete rows,
- $n' \in \mathbb{N}$ is the number of concrete columns, 
- the coordinate mapping function  
  $$
  \mathsf{coord\_map} : [0, m) \times [0, n) \to [0, m') \times [0, n')
  $$
   assigns concrete coordinates to each abstract cell position.

To translate the abstract circuit to a concrete circuit using the hints, we construct an injective mapping of abstract row numbers to concrete row numbers before applying offsets, such that:

> * all *constrained* abstract cells map to concrete cell coordinates that are in range;

Reviewer: I think I prefer to define it earlier, since it doesn't depend on the specific mapping we are defining here.

* every *constrained* abstract cell is represented by a distinct concrete cell, except that abstract cells that are equivalent under $\equiv$ *may* be identified.
  
### Hints

> The domain $\mathsf{Hints}$ represents collections of designer-provided **offset hints**. Each hint assigns to a row index $i \in [0, m)$:

Reviewer 2: Replace "row index" with "column index"
